File: cpp/0012-integer-to-roman/1.cpp
class Solution {
public:
    string intToRoman(int num) {
        const static struct {int num; char str[3];} m[13] = {
            {1000, "M"},
            {900, "CM"},
            {500, "D"},
            {400, "CD"},
            {100, "C"},
            {90, "XC"},
            {50, "L"},
            {40, "XL"},
            {10, "X"},
            {9, "IX"},
            {5, "V"},
            {4, "IV"},
            {1, "I"},
        };
        string res;
        
        for (auto &it : m) {
            while (num >= it.num) {
                num -= it.num;
                res += it.str;
            }
            if (num == 0) break;
        }
        return res;
    }
};

File: rust/0017-letter-combinations-of-a-phone-number/1.rs
const b: [&[u8]; 10] = [b"",b"",b"abc",b"def",b"ghi",b"jkl",b"mno", b"pqrs", b"tuv", b"wxyz"];
fn letter_combinations_impl(digits: &[u8]) -> Vec<String> {
    let l = digits.len();
    if l == 0 {return vec![];}

    let left = &digits[0..l-1];
    let left = letter_combinations_impl(left);
    let right = (digits[l-1] - b'0') as usize;

    if l == 1 {
        return Vec::from(b[right]).iter().map(|&num| {
            let mut ret = String::new();
            ret.push(num as char);
            ret
        }).collect();
    }
    
    let mut ret = Vec::new();
    for &c in b[right].iter() {
        for s in left.iter() {
            let mut o = s.to_owned();
            o.push(c as char);
            ret.push(o)
        }
    }
    ret
}

impl Solution {
    pub fn letter_combinations(digits: String) -> Vec<String> {
        letter_combinations_impl(digits.as_bytes())
    }
}
